crates/biome_service/src/scanner.rs

```diff
@@ -0,0 +1,211 @@
+use std::collections::BTreeSet;
+use std::panic::catch_unwind;
+use std::path::Path;
+use std::sync::{Once, RwLock};
+use std::thread;
+use std::time::{Duration, Instant};
+
+use crossbeam::channel::{unbounded, Receiver, Sender};
+use rayon::ThreadPoolBuilder;
+
+use biome_diagnostics::serde::Diagnostic;
+use biome_diagnostics::{Diagnostic as _, Error, Severity};
+use biome_fs::{BiomePath, PathInterner, TraversalContext, TraversalScope};
+
+use crate::workspace::{DocumentFileSource, OpenFileParams};
+use crate::{Workspace, WorkspaceError};
+
+pub(crate) struct ScanResult {
+    /// Diagnostics reported while scanning the project.
+    pub diagnostics: Vec<Diagnostic>,
+
+    /// Duration of the full scan.
+    pub duration: Duration,
+}
+
+pub(crate) fn scan(workspace: &dyn Workspace, folder: &Path) -> Result<ScanResult, WorkspaceError> {
+    init_thread_pool();
+
+    let (interner, _path_receiver) = PathInterner::new();
+    let (diagnostics_sender, diagnostics_receiver) = unbounded();
+
+    let collector = DiagnosticsCollector::new();
+
+    let (duration, diagnostics) = thread::scope(|scope| {
+        let handler = thread::Builder::new()
+            .name("biome::scanner".to_string())
+            .spawn_scoped(scope, || collector.run(diagnostics_receiver))
+            .expect("failed to spawn scanner thread");
+
+        // The traversal context is scoped to ensure all the channels it
+        // contains are properly closed once scanning finishes.
+        let (duration, _evaluated_paths) = scan_folder(
+            folder,
+            &ScanContext {
+                workspace,
+                interner,
+                diagnostics_sender,
+                evaluated_paths: Default::default(),
+            },
+        );
+
+        // Wait for the collector thread to finish.
+        let diagnostics = handler.join().unwrap();
+
+        (duration, diagnostics)
+    });
+
+    Ok(ScanResult {
+        diagnostics,
+        duration,
+    })
+}
+
+/// Sets up the global Rayon thread pool the first time it's called.
+///
+/// This is used to assign friendly debug names to the threads of the pool.
+fn init_thread_pool() {
+    static INIT_ONCE: Once = Once::new();
+    INIT_ONCE.call_once(|| {
+        ThreadPoolBuilder::new()
+            .thread_name(|index| format!("biome::workspace_worker_{index}"))
+            .build_global()
+            .expect("failed to initialize the global thread pool");
+    });
+}
+
+/// Initiates the filesystem traversal tasks from the provided path and runs it to completion.
+///
+/// Returns the duration of the process and the evaluated paths.
+fn scan_folder(folder: &Path, ctx: &ScanContext) -> (Duration, BTreeSet<BiomePath>) {
```

Comment:
Shouldn't we ensure that `folder` is actually a folder? In case the consumer passes a file